client/Wayland/wlf_cliprdr.c

```diff
@@ -0,0 +1,863 @@
+	wfClipboard* clipboard = (wfClipboard*) context->custom;
+	clipboard->streams_supported = FALSE;
+
+	for (i = 0; i < capabilities->cCapabilitiesSets; i++)
+	{
+		const CLIPRDR_CAPABILITY_SET* caps = (const CLIPRDR_CAPABILITY_SET*) capsPtr;
+
+		if (caps->capabilitySetType == CB_CAPSTYPE_GENERAL)
+		{
+			const CLIPRDR_GENERAL_CAPABILITY_SET* generalCaps = (const CLIPRDR_GENERAL_CAPABILITY_SET*) caps;
+
+			if (generalCaps->generalFlags & CB_STREAM_FILECLIP_ENABLED)
+			{
+				clipboard->streams_supported = TRUE;
+			}
+		}
+
+		capsPtr += caps->capabilitySetLength;
+	}
+
+	return CHANNEL_RC_OK;
+}
+
+static void wlf_cliprdr_transfer_data(UwacSeat* seat, void* context, const char* mime, int fd)
+{
+	wfClipboard* clipboard = (wfClipboard*)context;
+	size_t x;
+	WINPR_UNUSED(seat);
+	clipboard->responseMime = NULL;
+
+	for (x = 0; x < ARRAYSIZE(mime_html); x++)
+	{
+		const char* mime_cur = mime_html[x];
+
+		if (strcmp(mime_cur, mime) == 0)
+		{
+			clipboard->responseMime = mime_cur;
+			clipboard->responseFormat = ClipboardGetFormatId(clipboard->system, "HTML Format");
+			break;
+		}
+	}
+
+	for (x = 0; x < ARRAYSIZE(mime_text); x++)
+	{
+		const char* mime_cur = mime_text[x];
+
+		if (strcmp(mime_cur, mime) == 0)
+		{
+			clipboard->responseMime = mime_cur;
+			clipboard->responseFormat = CF_UNICODETEXT;
+			break;
+		}
+	}
+
+	for (x = 0; x < ARRAYSIZE(mime_image); x++)
+	{
+		const char* mime_cur = mime_image[x];
+
+		if (strcmp(mime_cur, mime) == 0)
+		{
+			clipboard->responseMime = mime_cur;
+			clipboard->responseFormat = CF_DIB;
+			break;
+		}
+	}
+
+	if (clipboard->responseMime != NULL)
+	{
+		clipboard->responseFile = fdopen(fd, "w");
+
+		if (clipboard->responseFile)
+			wlf_cliprdr_send_data_request(clipboard, clipboard->responseFormat);
+	}
+}
+
+static void wlf_cliprdr_cancel_data(UwacSeat* seat, void* context)
+{
+	WINPR_UNUSED(seat);
+	WINPR_UNUSED(context);
+}
+
+/**
+ * Called when the clipboard changes server side.
+ *
+ * Clear the local clipboard offer and replace it with a new one
+ * that announces the formats we get listed here.
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT wlf_cliprdr_server_format_list(CliprdrClientContext* context,
+        const CLIPRDR_FORMAT_LIST* formatList)
+{
+	UINT32 i;
+	wfClipboard* clipboard;
+	BOOL html = FALSE;
+	BOOL text = FALSE;
+	BOOL image = FALSE;
+
+	if (!context || !context->custom)
+		return ERROR_INVALID_PARAMETER;
+
+	clipboard = (wfClipboard*) context->custom;
+	wlf_cliprdr_free_server_formats(clipboard);
+
+	if (!(clipboard->serverFormats = (CLIPRDR_FORMAT*) calloc(
+	                                     formatList->numFormats, sizeof(CLIPRDR_FORMAT))))
+	{
+		WLog_Print(clipboard->log, WLOG_ERROR, "failed to allocate %"PRIuz" CLIPRDR_FORMAT structs",
+		           clipboard->numServerFormats);
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	clipboard->numServerFormats = formatList->numFormats;
+
+	if (!clipboard->seat)
```

Comment:
This looks so unlikely that we should print a message too when we hit that case.